src/stream/src/task/lru_manager.rs

```diff
@@ -0,0 +1,341 @@
+        Arc::new(Self {
+            watermark_epoch: Arc::new(0.into()),
+            total_memory_available_bytes: 0,
+            barrier_interval_ms: 0,
+        })
+    }
+
+    pub fn create_cache<K: Hash + Eq, V>(&self) -> ManagedLruCache<K, V> {
+        ManagedLruCache {
+            inner: LruCache::unbounded(),
+            watermark_epoch: self.watermark_epoch.clone(),
+        }
+    }
+
+    pub fn create_cache_with_hasher_in<K: Hash + Eq, V, S: BuildHasher, A: Clone + Allocator>(
+        &self,
+        hasher: S,
+        alloc: A,
+    ) -> ManagedLruCache<K, V, S, A> {
+        ManagedLruCache {
+            inner: LruCache::unbounded_with_hasher_in(hasher, alloc),
+            watermark_epoch: self.watermark_epoch.clone(),
+        }
+    }
+
+    pub fn create_cache_with_hasher<K: Hash + Eq, V, S: BuildHasher>(
+        &self,
+        hasher: S,
+    ) -> ManagedLruCache<K, V, S> {
+        ManagedLruCache {
+            inner: LruCache::unbounded_with_hasher(hasher),
+            watermark_epoch: self.watermark_epoch.clone(),
+        }
+    }
+
+    fn set_watermark_time(&self, time: u64) {
+        let epoch = Epoch::from_physical_time(time).0;
+        let watermark_epoch = self.watermark_epoch.as_ref();
+        watermark_epoch.store(epoch, Ordering::Relaxed);
+    }
+
+    pub async fn run(self: Arc<Self>) {
+        let mem_threshold_graceful =
+            (self.total_memory_available_bytes as f64 * Self::EVICTION_THRESHOLD_GRACEFUL) as usize;
+        let mem_threshold_aggressive = (self.total_memory_available_bytes as f64
+            * Self::EVICTION_THRESHOLD_AGGRESSIVE) as usize;
+
+        let mut watermark_time = Epoch::physical_now();
+        let mut last_total_bytes_used = 0;
+        let mut step = 0;
+
+        loop {
+            // Wait for a while to check if need eviction.
+            sleep(Duration::from_millis(self.barrier_interval_ms as u64)).await;
+
+            let stats = INSTRUMENTED_JEMALLOC.stats();
+            let cur_total_bytes_used = stats.bytes_allocated - stats.bytes_deallocated;
+
+            step = if cur_total_bytes_used < mem_threshold_graceful {
+                // Do not evict if the memory usage is lower than `mem_threshold_graceful`
+                0
+            } else if cur_total_bytes_used < mem_threshold_aggressive {
+                // Gracefully evict
+                if last_total_bytes_used < cur_total_bytes_used {
+                    1
+                } else {
+                    step + 1
+                }
+            } else if last_total_bytes_used < cur_total_bytes_used {
+                // Aggressively evict
+                if step == 0 {
+                    2
+                } else {
+                    step * 2
+                }
+            } else {
+                step
+            };
+
+            last_total_bytes_used = cur_total_bytes_used;
+            watermark_time += self.barrier_interval_ms as u64 * step;
```

Comment:
There's no guarantee that there's a new epoch every `barrier_interval` time. 🤣 If there's a "big" epoch, it may take a long time to evict the data from this epoch. Is it okay?

Reply:
BTW, I still suggest introducing a "sequence number" for the epoch, for those usages that do not care about the "timestamp" property of the epoch. I believe debugging will also be benefited from this.

Reply:
It might be a problem. We can optimize it by triggering eviction when receiving a data chunk, but IIRC it may not work correctly in hash agg?

Reply:
However, the eviction granularity is still an epoch, so we're not able to "evict partial data from an epoch". 😄

> IIRC it may not work correctly in hash agg?

Yes. Still need some refactor if we really want to trigger eviction in an epoch.